lib/granite/action.rb

```diff
@@ -50,20 +50,28 @@ def assign_attributes(attributes)
     prepend AssignAttributes
 
     handle_exception ActiveRecord::RecordInvalid do |e|
-      errors.messages.deep_merge!(e.record.errors.messages) do |_, this, other|
-        (this + other).uniq
-      end
+      merge_errors(e.record.errors)
     end
 
     handle_exception ActiveData::ValidationError do |e|
-      errors.messages.deep_merge!(e.model.errors.messages) do |_, this, other|
-        (this + other).uniq
-      end
+      merge_errors(e.model.errors)
     end
 
     handle_exception Granite::Action::ValidationError do |e|
-      errors.messages.deep_merge!(e.action.errors.messages) do |_, this, other|
-        (this + other).uniq
+      merge_errors(e.action.errors)
+    end
+
+    if ActiveModel.version < Gem::Version.new('6.1.0')
+      def merge_errors(other_errors)
+        errors.messages.deep_merge!(other_errors.messages) do |_, this, other|
+          (this + other).uniq
+        end
+      end
+    else
+      def merge_errors(other_errors)
+        other_errors.each do |error|
+          errors.import(error) unless errors.added?(error.attribute, error)
+        end
       end
     end
```

Comment:
This branching is the reason coverage has decreased. Only one of the two branches will ever be used in a single test run.

Anyway, the tests run in ruby 2.4 use activemodel 5, so the legacy execution branch is covered.